documentation/src/docs/security/securing-distributed-mode-cluster.md

```diff
@@ -0,0 +1,420 @@
+              -storepass <keystore-password> -keypass <key-password> \
+              -dname <distinguished-name> -ext SAN=DNS:<hostname>,
+
+  # Passwords and other arguments are entered interactively on the prompt in this case.
+  $ keytool -keystore server01.keystore.jks -alias server01 -genkey
+  ```
+* A weak password `changeit` is used everywhere, for easier reading. Be sure to replace it with a strong and separate
+password for each file.
+
+### Stage 1: Setting Up a Certificate Authority (CA)
+
+If you are going to use an existing public or internal CA service or certificate and key bundle, you may skip this part altogether, and go to [Obtaining Server Certificates and keys](#stage-2-obtaining-server-certificates-and-keys).
+
+Here, we'll generate a CA in the form of a public/private key pair and a self-signed certificate.
+Later, we'll use the CA certificate/key bundle to sign server certificates used in the cluster.
+
+1. Generate a certificate and public/private key pair, for use as a CA.
+
+   ```bash
+   # All inputs provided using command line arguments
+   $ openssl req -new -x509 -keyout ca-key -out ca-cert -days <validity> \
+            -subj "<distinguished_name>" \
+            -passout pass:<strong_password>
+
+   # Sample command
+   $ openssl req -new -x509 -keyout ca-key -out ca-cert -days 365 \
+            -subj "/C=US/ST=Washington/L=Seattle/O=Pravega/OU=CA/CN=Pravega-CA" \
+            -passout pass:changeit
+   ```
+
+2. Create a truststore containing the CA's certificate.
+
+   This truststore will be used by external and internal clients. External clients are client applications connected
+   to a Pravega cluster. Services running on server nodes play the role of internal clients when accessing other services.
+
+   ```bash
+   $ keytool -keystore client.truststore.jks -noprompt -alias CARoot -import -file ca-cert \
+        -storepass changeit
+
+   # Optionally, list the truststore's contents to verify everything is in order. The output should show
+   # a single entry with alias name `caroot` and entry type `trustedCertEntry`.
+   $ keytool -list -v -keystore client.truststore.jks -storepass changeit
+   ```
+
+At this point, the following CA and client truststore artifacts shall be  available:
+
+| File | Description |
+|:-----:|:--------|
+| `ca-cert` | PEM-encoded X.509 certificate of the CA |
+| `ca-key` | PEM-encoded file containing the CA's encrypted private key  |
+| `client.truststore.jks` | A password-protected truststore file containing the CA's certificate |
+
+### Stage 2: Obtaining Server Certificates and keys
+
+This stage is about performing the following steps for each service.
+
+1. Generating server certificates and keys
+2. Generating a Certificate Signing Request (CSR)
+3. Submitting the CSR to a CA and obtaining a signed certificate
+4. Preparing a keystore containing the signed server certificate and the CA's certificate
+5. Exporting the server certificate's private key
+
+NOTE:
+
```

Comment:
**Note:** For services running on the same host, the same certificate can be used, if those services are accessed using the same hostname/IP address. Also, wildcard certificates can be used to share certificates across hosts. However, it is strongly recommended that separate certificates be used for each service.

Reply:
Done.